// core-marcenaria/connectors/mongodb/schemas/analytical.js
const mongoose = require('mongoose')

module.exports.AnalyticalSchema = new mongoose.Schema({
  projectId: { type: Object, required: true },
  campaignId: { type: Object },
  keyAccess: { type: String, required: true },
  issueDate: { type: Date },

  invoiceId: { type: Object, required: true },
  invoiceCreatedAt: { type: Date, required: true },
  invoiceTotalValue: { type: Number },

  userId: { type: Object, required: true },
  userFirstName: { type: String, default: '' },
  userLastName: { type: String, default: '' },
  userFullName: { type: String, default: '' },
  userCpf: { type: String, default: null },
  userOccupation: { type: String, default: null },
  userEmail: { type: String, default: null },
  userGender: { type: String, default: null },
  userState: { type: String, default: null },
  userCity: { type: String, default: null },
  userCompany: { type: String, default: null },
  userStateCompany: { type: String, default: null },
  userCityCompany: { type: String, default: null },
  userBirthday: { type: Date },
  userCreateAt: { type: Date },

  issuerName: { type: String, default: null },
  issuerDocumentNumber: { type: String, default: null },
  issuerDocumentType: { type: String, enum: ['CPF', 'CNPJ'] },
  issuerAddressCity: { type: String, default: null },
  issuerAddressCountry: { type: String, default: null },
  issuerAddressStateIso: { type: String, default: null },

  recipientName: { type: String, default: null },
  recipientDocumentNumber: { type: String, default: null },
  recipientDocumentType: { type: String, enum: ['CPF', 'CNPJ'] },
  recipientAddressCity: { type: String, default: null },
  recipientAddressCountry: { type: String, default: null },
  recipientAddressStateIso: { type: String, default: null },

	status: { type: String, default: null },
	classification: { type: String, default: null },
  line: { type: String, default: null },
  pattern: { type: String, default: null },
  height: { type: String, default: null },
  width: { type: String, default: null },
  thickness: { type: String, default: null },
  properties: { type: String, default: null },

  name: { type: String, default: null },
  description: { type: String, default: null },
  itemNumber: { type: Number, default: null },
  commercialUnit: { type: String, default: null },
  commercialQuantity: { type: Number, default: null },
  commercialValue: { type: Number, default: null },
  tributaryUnit: { type: String, default: null },
  tributaryQuantity: { type: Number, default: null },
  tributaryValue: { type: Number, default: null },
  value: { type: Number, default: null },

  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
  sentToS3: { type: Boolean, default: false }
})
