// core-marcenaria/connectors/mongodb/schemas/transaction.js
const mongoose = require('mongoose')

module.exports.STATUS_COMPETITOR = 'competitor'
module.exports.STATUS_INVALID = 'invalid'
module.exports.STATUS_LIMIT = 'exceeded limit'
module.exports.STATUS_APPROVED = 'approved'
module.exports.STATUS_PARCIAL_APPROVED = 'parcial'
module.exports.STATUS_BLACKLIST = 'blacklist'

module.exports.CLASSIFICATION_DURATEX = 'duratex'

module.exports.TransactionSchema = new mongoose.Schema({
  userId: mongoose.Schema.Types.ObjectId,
  adminUserId: mongoose.Schema.Types.ObjectId,
  invoiceId: mongoose.Schema.Types.ObjectId,
  campaignId: mongoose.Schema.Types.ObjectId,
  description: String,
  baseValue: Number,
  totalValue: Number,
  creditedValue: Number,
  quantity: Number,
  status: String,
  classificationProblems: Object,
  cronClassificationErrors: { type: Boolean, default: false },
  pending: { type: Boolean, default: true },
  classification: String,
  line: String,
  pattern: String,
  height: { type: Number, default: 0 },
  width: { type: Number, default: 0 },
  thickness: { type: Number, default: 0 },
  createdAt: { type: Date, default: new Date() },
  updatedAt: { type: Date, default: new Date() }
})
