// core-marcenaria/connectors/mongodb/schemas/user-club.js
const mongoose = require('mongoose')

module.exports.guerreirosZSchema = new mongoose.Schema({
  "date":{ type: Date, default: Date.now },
  "nome":{ type: String },
  "planetaDeOrigem":{ type: String },
  "raça": { type: String },
  "sexo": { type: String },
  "foiMorto":{ type: Boolean }
  
})


// module.exports.STATUS_CREATED = 1
// module.exports.STATUS_VALIDATED = 2
// module.exports.STATUS_RESETCONFIRMATION = 3
// module.exports.STATUS_RESETPASSWORD = 4

// module.exports.UserClubSchema = new mongoose.Schema({
//   firstName: { type: String },
//   lastName: { type: String },
// 	email: { type: String, lowercase: true },
// 	secondaryEmail: { type: String, lowercase: true },
//   avatar: { type: String },
//   login: { type: String, lowercase: true },
//   company: { type: String, default: 'plt-marceneiro' },
//   cpf: { type: String },
//   occupationSlug: { type: String },
//   validated: { type: Boolean, default: false },
// 	confirmationCode: { type: String },
// 	lastLogin: { type: Date },
//   status: {
//     type: Number,
//     enum: [
// 			this.STATUS_CREATED,
// 			this.STATUS_VALIDATED,
// 			this.STATUS_RESETCONFIRMATION,
// 			this.STATUS_RESETPASSWORD
// 		],
//     default: this.STATUS_CREATED
//   },
//   dataAuth: { type: Object },
//   createdAt: { type: Date, default: Date.now },
//   updatedAt: { type: Date, default: Date.now },
//   aboutMe: {
//     resume: { type: String },
//     from: { type: String },
//     gender: { type: String },
//     birthday: { type: Date },
//     address: { type: String },
//     numberAddress: { type: String },
//     neighborhood: { type: String },
//     complement: { type: String },
//     state: { type: String },
//     city: { type: String }
//   },
//   emailRememberToActivationAccount5days: { type: Boolean, default: false },
//   emailRememberToActivationAccount20days: { type: Boolean, default: false },
//   contact: {
//     email: { type: String },
//     telephone: { type: String },
//     cep: { type: String }
//   },
//   aboutUs: {
//     name: { type: String },
//     cnpj: { type: String },
//     email: { type: String },
//     telephone: { type: String },
//     cep: { type: String },
//     address: { type: String },
//     numberAddress: { type: String },
//     neighborhood: { type: String },
//     complement: { type: String },
//     state: { type: String },
//     city: { type: String },
//     site: { type: String }
//   },
//   partners: {
//     ipremi: {
//       userIdExternal: { type: Number }
//     }
//   }
// })
