// core-marcenaria/integrations/notification.js
const axios = require('axios')

class NotificationClass {
	constructor() {
		const {
			NOTIFICATION_API_URL,
			NOTIFICATION_API_XAPIKEY
		} = process.env
		
		this.axiosInstance = axios.create({
			baseURL: NOTIFICATION_API_URL,
			headers: { 'x-api-key': NOTIFICATION_API_XAPIKEY}
		})
	}

	async sendWoodWorkerInvoiceApproved(data) {
		await this.axiosInstance.post('/email/invoice/approved', data)
	}

	async sendWoodWorkerInvoiceRefused(data) {
		await this.axiosInstance.post('/email/invoice/refused', data)
	}

	async sendCongrats (data) {
		await this.axiosInstance.post('/email/congratulations/dura-coin', data)
	}

	async accountActivation (data) {
		await this.axiosInstance.post('/email/account/activation', data)
	}
	async createUser (data) {
		data.subject = 'Clube Duratex - Ativação de Conta'
		await this.axiosInstance.post('/email/user/create', data)
	}

	async sendExpiredDuracoin (data) {
		await this.axiosInstance.post('/email/expired/dura-coin', data)
	}

	async forgotPassword (data) {
		await this.axiosInstance.post('/email/user/forgot-password', data)
	}

	async welcome(data) {
		await this.axiosInstance.post('/email/user/welcome', data)
	}

	async resendEmailValidation(data) {
		await this.axiosInstance.post('/email/user/resend-validation', data)
	}

}

module.exports = new NotificationClass()
